javascript/ql/src/semmle/javascript/ApiGraphs.qll

```diff
@@ -738,25 +738,27 @@ module API {
         boundArgs in [0 .. 10]
       )
       or
-      exists(StepSummary summary |
-        t = useStep(nd, promisified, boundArgs, result, summary).append(summary)
-      )
+      t = useStep(nd, promisified, boundArgs, result)
     }
 
     private import semmle.javascript.dataflow.internal.StepSummary
 
     /**
      * Holds if `nd`, which is a use of an API-graph node, flows in zero or more potentially
      * inter-procedural steps to some intermediate node, and then from that intermediate node to
-     * `res` in one step described by `summary`.
+     * `res` in one step described by the resulting TypeTracker.
      *
      * This predicate exists solely to enforce a better join order in `trackUseNode` above.
      */
-    pragma[noinline]
+    pragma[noopt]
```

Comment:
Can you comment on what other optimisations besides inlining you had to prevent here? (As you know, `noopt` is a bit of a double-edged sword, so I want to make sure we have thought carefully about using it.)

Reply:
I need a specific join-order, and I need that specific join-order inside a single predicate. 
(I think, I'm not sure how the runtime does things, but having a large temporary result seems to run fast, while having a predicate that has a large result is slow).  

Here are tuple counts for the `pragma[noopt]` predicate: 

```Text
4363     ~8%     {5} r1 = SCAN ApiGraphs::API::Impl::trackUseNode#2#fffff#prev_delta OUTPUT In.4, In.0 'nd', In.1 'promisified', In.2 'boundArgs', In.3
11079952 ~3%     {6} r2 = JOIN r1 WITH StepSummary::StepSummary::step#fff@staged_ext ON FIRST 1 OUTPUT Lhs.4, Rhs.2, Lhs.1 'nd', Lhs.2 'promisified', Lhs.3 'boundArgs', Rhs.1 'res'
19536    ~4%     {5} r3 = JOIN r2 WITH TypeTracking::TypeTracker::append_dispred#fff@staged_ext ON FIRST 2 OUTPUT Lhs.2 'nd', Lhs.3 'promisified', Lhs.4 'boundArgs', Lhs.5 'res', Rhs.2 'result'
19476    ~4%     {5} r4 = r3 AND NOT ApiGraphs::API::Impl::useStep#fffff#prev(Lhs.0 'nd', Lhs.1 'promisified', Lhs.2 'boundArgs', Lhs.3 'res', Lhs.4 'result')
                 return r4
```

The predicate is pretty fast even though `JOIN r1 WITH StepSummary::StepSummary::step` is large.   

I just couldn't get that specific join-order, in a single predicate, without using `noopt`. 
I can get the join-order I want if I split up the predicate, but then I get something like the below (in a helper predicate), which runs about 10X slower on `ol2` (2s vs 20s). 
```
Tuple counts for ApiGraphs::API::Impl::useStepHelper#ffffff/6@i50#c9455c:
4340     ~7%     {5} r1 = SCAN ApiGraphs::API::Impl::trackUseNode#2#fffff#prev_delta OUTPUT In.4, In.0 'nd', In.1 'promisified', In.2 'boundArgs', In.3 'result'
11079929 ~2%     {6} r2 = JOIN r1 WITH StepSummary::StepSummary::step#fff@staged_ext ON FIRST 1 OUTPUT Lhs.1 'nd', Lhs.2 'promisified', Lhs.3 'boundArgs', Lhs.4 'result', Rhs.1 'res', Rhs.2 'summary'
11079921 ~2%     {6} r3 = r2 AND NOT ApiGraphs::API::Impl::useStepHelper#ffffff#prev(Lhs.0 'nd', Lhs.1 'promisified', Lhs.2 'boundArgs', Lhs.4 'res', Lhs.5 'summary', Lhs.3 'result')
11079921 ~4%     {6} r4 = SCAN r3 OUTPUT In.0 'nd', In.1 'promisified', In.2 'boundArgs', In.4 'res', In.5 'summary', In.3 'result'
                 return r4
```